extensions/functions_arithmetic.yaml

```diff
@@ -173,6 +173,178 @@ scalar_functions:
           - value: i64
           - value: i64
         return: i64
+  -
+    name: "cos"
+    description: "Get the cosine of a value in radians."
+    impls:
+      - args:
+          - value: i8
```

Comment:
Adding to this, in Arrow-c++, we only have kernels for fp32 and fp64 (I believe the docs may be wrong here and imply support for decimal inputs but we insert an implicit cast).  You won't get an error if you try the integer types but it will insert an implicit cast.

Reply:
I only included all of the others to follow how some other functions have been specified, but I don't see any value in doing it this way if we don't need to.

Reply:
I think, if the end result is no different than a cast the kernel is not needed.  If there are other cases like this we should get rid of those kernels as well.

Reply:
Perhaps we should allow an impl here to specify multiple allowed types for a single value arg, e.g.
```yaml
    impls:
      - args:
           - value: [ i8, i16, i32, i64, fp32,  fp64 ]
         return: fp64
```
That would reduce the amount of boilerplate in these YAML files and at the same time avoid Substrait producers needing to insert lots of fussy explicit casts.

Reply:
@ianmcook we explored patterns around this [early](https://github.com/substrait-io/substrait/commit/9e84da55393a24953ed9f9869fa423f86f4860f7) but it became a slippery slope of complexity for implementers. Ultimately, we figured that since yamls aren't something that will be worked with super-frequently it was ok to be a bit more verbose to simplify implementation of consumption of the yaml.

Reply:
>Adding to this, in Arrow-c++, we only have kernels for fp32 and fp64 (I believe the docs may be wrong here and imply support for decimal inputs but we insert an implicit cast). You won't get an error if you try the integer types but it will insert an implicit cast.

@westonpace Is that correct? In an Acero ExecPlan, I believe `add_checked(int8, int8)` returns `int8`

Reply:
@jacques-n that makes sense, but this seems to create a dilemma in which we must decide to do one or the other of these things:

1. Have producers include many explicit casts in Substrait plans (even when they might not strictly be necessary) which I believe adds computational cost for engines executing these plans
2. Add impls to these YAML files covering every combination of input types and the associated output type, which creates a minor combinatorial explosion even just for the binary functions

Reply:
A few thoughts:

* I don't understand why the explicit casts would cost extra compute if they are already being done implicitly. 
* I agree with Weston's comment: "If there are other cases like this we should get rid of those kernels as well." 
* From the Substrait perspective: 2 is the "right" approach. Define the specific things that are possible. Avoid trying to be overly fancy in the function definition paradigm. Use a script to define things if you want. (This isn't a place we need optimization, generally.)

Reply:
To clarify what I mean in the comment above: consider the case of `add(i32, i64)`. In every engine of which I am aware, that returns `i64`. If you reverse the order of the value args so it's `add(i64, i32)`, that also returns `i64`. So should we add impls for `add` like so?
```yaml
      - args:
          - options: [ SILENT, SATURATE, ERROR ]
            required: false
          - value: i32
          - value: i64
        return: i64
      - args:
          - options: [ SILENT, SATURATE, ERROR ]
            required: false
          - value: i64
          - value: i32
        return: i64
```
and repeat this for every permutation of input for which engines tend to behave the same way? I read through https://github.com/substrait-io/substrait/issues/65 and it all seems reasonable but does not speak to this question.

Reply:
Yes, the current pattern requires this. 

Do you have an i32 + i64? Is it because of some performance benefit? I would expect that to be rare. The most common pattern systems have is an implementation of i64 + i64 and they cast i32 to i64 before doing i64 + i64.

Reply:
I've updated to just support i64 and fp64 now.

Reply:
I think `fp32` is probably more valuable than `i64` and I apologize for being obnoxious :).  My goal isn't to limit the number of kernels but to create one kernel per distinct implementation.  I think it is common for hardware/engines to have different implementations for float and double.  By "different implementation" I mean "passing the same inputs yields a different result"

For an example, consider the [C++ standard library implementation for cosine](https://en.cppreference.com/w/cpp/numeric/math/cos).  There are three ways to call `std::cos`, with a `float`, with a `double`, or with an `integral` type.  However, calling cosine with an integral type is identical to casting to double and calling cosine with that double value.  We can see this both in the docs:

> 4) A set of overloads or a function template accepting an argument of any [integral type](https://en.cppreference.com/w/cpp/types/is_integral). Equivalent to 2) (the argument is cast to double).

and in practice:

https://godbolt.org/z/h6Pb5WY86

Reply:
Thanks for going into detail about that, makes sense!  Updated to now have `fp32 ` and `fp64`

Reply:
Part of this discussion is somewhat related to https://github.com/substrait-io/substrait/issues/251, so linking for posterity.